File: CS/CustomAppointmentFlyoutExample/Form1.cs
using DevExpress.Utils.Drawing;
using DevExpress.Utils.Svg;
using DevExpress.XtraEditors;
using DevExpress.XtraScheduler;
using System;
using System.Drawing;
using System.IO;
using System.Reflection;

namespace CustomAppointmentFlyoutExample {
    public partial class Form1 : XtraForm {
        readonly FontStorage fontStorage;

        public Form1() {
            this.fontStorage = new FontStorage();
            InitializeComponent();
            InitializeScheduler();
        }

        void InitializeScheduler() {
            DataHelper.InitResources(this.schedulerStorage1);
            DataHelper.InitAppointments(this.schedulerStorage1);
            this.schedulerControl1.Start = DateTime.Today;
            this.schedulerControl1.GroupType = SchedulerGroupType.Resource;
            #region #OptionsFlyout
            schedulerControl1.OptionsFlyout.SubjectAutoHeight = true;
            #endregion #OptionsFlyout
        }

        #region #CustomizeAppointmentFlyout
        void OnSchedulerControlCustomizeAppointmentFlyout(object sender, CustomizeAppointmentFlyoutEventArgs e) {
            e.ShowEndDate = false;
            e.ShowReminder = false;
            e.ShowLocation = true;
            e.SubjectAppearance.Font = fontStorage.SubjectAppearanceFont;
            e.Location = "N/A";
        }
        #endregion #CustomizeAppointmentFlyout

        #region #CustomDrawAppointmentFlyoutSubject
        void OnSchedulerControlCustomDrawAppointmentFlyoutSubject(object sender, CustomDrawAppointmentFlyoutSubjectEventArgs e) {
            e.Handled = true;
            CustomDrawAppointmentFlyoutSubject(e);
        }
        void CustomDrawAppointmentFlyoutSubject(CustomDrawAppointmentFlyoutSubjectEventArgs e) {
            AppointmentBandDrawerViewInfoBase viewInfo = (AppointmentBandDrawerViewInfoBase)e.ObjectInfo;
            e.DrawBackgroundDefault();
            CustomDrawAppointmentFlyoutSubject(e.Appointment, viewInfo);
        }
        void CustomDrawAppointmentFlyoutSubject(Appointment appointment, AppointmentBandDrawerViewInfoBase viewInfo) {
            GraphicsCache cache = viewInfo.Cache;
            StringFormat stringFormat = new StringFormat(viewInfo.View.Appearance.GetStringFormat());
            stringFormat.Alignment = stringFormat.LineAlignment = StringAlignment.Center;
            try {
                // Draw status
                Rectangle statusRect = GetStatusBounds(viewInfo);
                cache.FillRectangle(viewInfo.View.Status.GetBrush(), statusRect);

                if (viewInfo.View.Status.Type == AppointmentStatusType.Free) {
                    // Draw a warning
                    cache.DrawImage(GetWarningIcon(new Size(statusRect.Height, statusRect.Height)), statusRect.Location);
                    cache.DrawString("Status is unacceptable", fontStorage.StatusFont, Brushes.Red, statusRect, stringFormat);
                }
                // Draw subject                        
                cache.DrawString(appointment.Subject, fontStorage.SubjectFont, Brushes.Black, GetSubjectBounds(viewInfo), stringFormat);
            }
            finally {
                stringFormat.Dispose();
            }
        }
        Rectangle GetSubjectBounds(AppointmentBandDrawerViewInfoBase viewInfo) {
            Rectangle bounds = viewInfo.Bounds;
            if (viewInfo.View.Status.Type == AppointmentStatusType.Free) {
                bounds.Offset(0, 10);
            }
            return bounds;
        }
        Rectangle GetStatusBounds(AppointmentBandDrawerViewInfoBase viewInfo) {
            Rectangle bounds = Rectangle.Inflate(viewInfo.Bounds, -1, -1);
            if (viewInfo.View.Status.Type == AppointmentStatusType.Free) {
                bounds.Height = 20;
            }
            else {
                bounds.Height = 5;
            }
            return bounds;
        }
        Image GetWarningIcon(Size size) {
            using (Stream stream = AppAssembly.GetManifestResourceStream("CustomAppointmentFlyoutExample.warning.svg")) {
                var paletteProvider = SvgPaletteHelper.GetSvgPalette(schedulerControl1.LookAndFeel, ObjectState.Selected);
                return SvgBitmap.FromStream(stream).Render(size, paletteProvider);
            }
        }
        static readonly Assembly AppAssembly = Assembly.GetExecutingAssembly();
        #endregion #CustomDrawAppointmentFlyoutSubject

        #region #AppointmentFlyoutShowing
        void OnSchedulerControl1AppointmentFlyoutShowing(object sender, AppointmentFlyoutShowingEventArgs e) {
            e.Control = new MyFlyout(e.FlyoutData.Subject, e.FlyoutData.Start, e.FlyoutData.End);
        }
        #endregion #AppointmentFlyoutShowing

        private void chkBtnTooltips_CheckedChanged(object sender, EventArgs e) {
            if (chkBtnTooltips.Checked) {
                #region #AllowDisplayAppointmentFlyout
                schedulerControl1.OptionsCustomization.AllowDisplayAppointmentFlyout = false;
                schedulerControl1.OptionsView.ToolTipVisibility = ToolTipVisibility.Always;
                #endregion #AllowDisplayAppointmentFlyout
            }
            else {
                schedulerControl1.OptionsCustomization.AllowDisplayAppointmentFlyout = true;
                schedulerControl1.OptionsView.ToolTipVisibility = ToolTipVisibility.Never;
            }
        }
        private void chkBtnAppointmentFlyoutShowing_CheckedChanged(object sender, EventArgs e) {
            if (chkBtnAppointmentFlyoutShowing.Checked)
            schedulerControl1.AppointmentFlyoutShowing += OnSchedulerControl1AppointmentFlyoutShowing;
            else
                schedulerControl1.AppointmentFlyoutShowing -= OnSchedulerControl1AppointmentFlyoutShowing;
        }


        private void chkBtnCustomizeAppointmentFlyout_CheckedChanged(object sender, EventArgs e) {
            if (chkBtnCustomizeAppointmentFlyout.Checked)
                schedulerControl1.CustomizeAppointmentFlyout += OnSchedulerControlCustomizeAppointmentFlyout;
            else
                schedulerControl1.CustomizeAppointmentFlyout -= OnSchedulerControlCustomizeAppointmentFlyout;
        }

        private void chkBtnCustomDrawAppointmentFlyoutSubject_CheckedChanged(object sender, EventArgs e) {
            if (chkBtnCustomDrawAppointmentFlyoutSubject.Checked)
                schedulerControl1.CustomDrawAppointmentFlyoutSubject += OnSchedulerControlCustomDrawAppointmentFlyoutSubject;
            else
                schedulerControl1.CustomDrawAppointmentFlyoutSubject -= OnSchedulerControlCustomDrawAppointmentFlyoutSubject;
        }
    }

    public class FontStorage : IDisposable {
        readonly Font subjectFont;
        readonly Font statusFont;
        readonly Font subjectAppearanceFont;

        public FontStorage() {
            this.subjectFont = new Font("Tahoma", 10, FontStyle.Bold);
            this.statusFont = new Font("Tahoma", 8, FontStyle.Bold);
            this.subjectAppearanceFont = new Font("Arial", 24);
        }

        public void Dispose() {
            Dispose(true);
        }
        void Dispose(bool disposing) {
            if (disposing) {
                SubjectFont.Dispose();
                StatusFont.Dispose();
                SubjectAppearanceFont.Dispose();
            }
        }

        public Font SubjectFont { get { return subjectFont; } }
        public Font StatusFont { get { return statusFont; } }
        public Font SubjectAppearanceFont { get { return subjectAppearanceFont; } }
    }
}

File: VB/CustomAppointmentFlyoutExample/Form1.vb
Imports DevExpress.Utils.Drawing
Imports DevExpress.Utils.Svg
Imports DevExpress.XtraEditors
Imports DevExpress.XtraScheduler
Imports System
Imports System.Drawing
Imports System.IO
Imports System.Reflection

Namespace CustomAppointmentFlyoutExample

    Public Partial Class Form1
        Inherits XtraForm

        Private ReadOnly fontStorage As FontStorage

        Public Sub New()
            fontStorage = New FontStorage()
            InitializeComponent()
            InitializeScheduler()
        End Sub

        Private Sub InitializeScheduler()
            InitResources(schedulerStorage1)
            InitAppointments(schedulerStorage1)
            schedulerControl1.Start = Date.Today
            schedulerControl1.GroupType = SchedulerGroupType.Resource
#Region "#OptionsFlyout"
            schedulerControl1.OptionsFlyout.SubjectAutoHeight = True
#End Region  ' #OptionsFlyout
        End Sub

#Region "#CustomizeAppointmentFlyout"
        Private Sub OnSchedulerControlCustomizeAppointmentFlyout(ByVal sender As Object, ByVal e As CustomizeAppointmentFlyoutEventArgs)
            e.ShowEndDate = False
            e.ShowReminder = False
            e.ShowLocation = True
            e.SubjectAppearance.Font = fontStorage.SubjectAppearanceFont
            e.Location = "N/A"
        End Sub

#End Region  ' #CustomizeAppointmentFlyout
#Region "#CustomDrawAppointmentFlyoutSubject"
        Private Sub OnSchedulerControlCustomDrawAppointmentFlyoutSubject(ByVal sender As Object, ByVal e As CustomDrawAppointmentFlyoutSubjectEventArgs)
            e.Handled = True
            CustomDrawAppointmentFlyoutSubject(e)
        End Sub

        Private Sub CustomDrawAppointmentFlyoutSubject(ByVal e As CustomDrawAppointmentFlyoutSubjectEventArgs)
            Dim viewInfo As AppointmentBandDrawerViewInfoBase = CType(e.ObjectInfo, AppointmentBandDrawerViewInfoBase)
            e.DrawBackgroundDefault()
            CustomDrawAppointmentFlyoutSubject(e.Appointment, viewInfo)
        End Sub

        Private Sub CustomDrawAppointmentFlyoutSubject(ByVal appointment As Appointment, ByVal viewInfo As AppointmentBandDrawerViewInfoBase)
            Dim cache As GraphicsCache = viewInfo.Cache
            Dim stringFormat As StringFormat = New StringFormat(viewInfo.View.Appearance.GetStringFormat())
            stringFormat.LineAlignment = StringAlignment.Center
            stringFormat.Alignment = stringFormat.LineAlignment
            Try
                ' Draw status
                Dim statusRect As Rectangle = GetStatusBounds(viewInfo)
                cache.FillRectangle(viewInfo.View.Status.GetBrush(), statusRect)
                If viewInfo.View.Status.Type = AppointmentStatusType.Free Then
                    ' Draw a warning
                    cache.DrawImage(GetWarningIcon(New Size(statusRect.Height, statusRect.Height)), statusRect.Location)
                    cache.DrawString("Status is unacceptable", fontStorage.StatusFont, Brushes.Red, statusRect, stringFormat)
                End If

                ' Draw subject                        
                cache.DrawString(appointment.Subject, fontStorage.SubjectFont, Brushes.Black, GetSubjectBounds(viewInfo), stringFormat)
            Finally
                stringFormat.Dispose()
            End Try
        End Sub

        Private Function GetSubjectBounds(ByVal viewInfo As AppointmentBandDrawerViewInfoBase) As Rectangle
            Dim bounds As Rectangle = viewInfo.Bounds
            If viewInfo.View.Status.Type = AppointmentStatusType.Free Then
                bounds.Offset(0, 10)
            End If

            Return bounds
        End Function

        Private Function GetStatusBounds(ByVal viewInfo As AppointmentBandDrawerViewInfoBase) As Rectangle
            Dim bounds As Rectangle = Rectangle.Inflate(viewInfo.Bounds, -1, -1)
            If viewInfo.View.Status.Type = AppointmentStatusType.Free Then
                bounds.Height = 20
            Else
                bounds.Height = 5
            End If

            Return bounds
        End Function

        Private Function GetWarningIcon(ByVal size As Size) As Image
            Using stream As Stream = AppAssembly.GetManifestResourceStream("CustomAppointmentFlyoutExample.warning.svg")
                Dim paletteProvider = SvgPaletteHelper.GetSvgPalette(schedulerControl1.LookAndFeel, ObjectState.Selected)
                Return SvgBitmap.FromStream(stream).Render(size, paletteProvider)
            End Using
        End Function

        Private Shared ReadOnly AppAssembly As Assembly = Assembly.GetExecutingAssembly()

#End Region  ' #CustomDrawAppointmentFlyoutSubject
#Region "#AppointmentFlyoutShowing"
        Private Sub OnSchedulerControl1AppointmentFlyoutShowing(ByVal sender As Object, ByVal e As AppointmentFlyoutShowingEventArgs)
            e.Control = New MyFlyout(e.FlyoutData.Subject, e.FlyoutData.Start, e.FlyoutData.End)
        End Sub

#End Region  ' #AppointmentFlyoutShowing
        Private Sub chkBtnTooltips_CheckedChanged(ByVal sender As Object, ByVal e As EventArgs)
            If chkBtnTooltips.Checked Then
#Region "#AllowDisplayAppointmentFlyout"
                schedulerControl1.OptionsCustomization.AllowDisplayAppointmentFlyout = False
#End Region  ' #AllowDisplayAppointmentFlyout
                schedulerControl1.OptionsView.ToolTipVisibility = ToolTipVisibility.Always
            Else
                schedulerControl1.OptionsCustomization.AllowDisplayAppointmentFlyout = True
                schedulerControl1.OptionsView.ToolTipVisibility = ToolTipVisibility.Never
            End If
        End Sub

        Private Sub chkBtnAppointmentFlyoutShowing_CheckedChanged(ByVal sender As Object, ByVal e As EventArgs)
            If chkBtnAppointmentFlyoutShowing.Checked Then
                AddHandler schedulerControl1.AppointmentFlyoutShowing, AddressOf OnSchedulerControl1AppointmentFlyoutShowing
            Else
                RemoveHandler schedulerControl1.AppointmentFlyoutShowing, AddressOf OnSchedulerControl1AppointmentFlyoutShowing
            End If
        End Sub

        Private Sub chkBtnCustomizeAppointmentFlyout_CheckedChanged(ByVal sender As Object, ByVal e As EventArgs)
            If chkBtnCustomizeAppointmentFlyout.Checked Then
                AddHandler schedulerControl1.CustomizeAppointmentFlyout, AddressOf OnSchedulerControlCustomizeAppointmentFlyout
            Else
                RemoveHandler schedulerControl1.CustomizeAppointmentFlyout, AddressOf OnSchedulerControlCustomizeAppointmentFlyout
            End If
        End Sub

        Private Sub chkBtnCustomDrawAppointmentFlyoutSubject_CheckedChanged(ByVal sender As Object, ByVal e As EventArgs)
            If chkBtnCustomDrawAppointmentFlyoutSubject.Checked Then
                AddHandler schedulerControl1.CustomDrawAppointmentFlyoutSubject, AddressOf OnSchedulerControlCustomDrawAppointmentFlyoutSubject
            Else
                RemoveHandler schedulerControl1.CustomDrawAppointmentFlyoutSubject, AddressOf OnSchedulerControlCustomDrawAppointmentFlyoutSubject
            End If
        End Sub
    End Class

    Public Class FontStorage
        Implements IDisposable

        Private ReadOnly subjectFontField As Font

        Private ReadOnly statusFontField As Font

        Private ReadOnly subjectAppearanceFontField As Font

        Public Sub New()
            subjectFontField = New Font("Tahoma", 10, FontStyle.Bold)
            statusFontField = New Font("Tahoma", 8, FontStyle.Bold)
            subjectAppearanceFontField = New Font("Arial", 24)
        End Sub

        Public Sub Dispose() Implements IDisposable.Dispose
            Dispose(True)
        End Sub

        Private Sub Dispose(ByVal disposing As Boolean)
            If disposing Then
                SubjectFont.Dispose()
                StatusFont.Dispose()
                SubjectAppearanceFont.Dispose()
            End If
        End Sub

        Public ReadOnly Property SubjectFont As Font
            Get
                Return subjectFontField
            End Get
        End Property

        Public ReadOnly Property StatusFont As Font
            Get
                Return statusFontField
            End Get
        End Property

        Public ReadOnly Property SubjectAppearanceFont As Font
            Get
                Return subjectAppearanceFontField
            End Get
        End Property
    End Class
End Namespace
